### PR TITLE
🎉 New Source: Oura

### PR DESCRIPTION
**Pull request by isaacharrisholt**
## What

Some individuals will want ot use Airbyte to collate their personal data for analysis, such as health and well-being data. One source of such data is Oura, a company that makes fitness wearables in the form of rings that provide scores on daily activity, readiness, and sleep, as well as heart rate data, workout data, and so on.

## How

This PR uses the lowcode CDK to create a GoCardless connector with the following streams:

- `daily_activity`
- `daily_readiness`
- `daily_sleep`
- `heart_rate`
- `sessions`
- `sleep_periods`
- `tags`
- `workouts`

These are all the endpoints currently available in Oura's v2 API except the personal info endpoint.

## Recommended reading order

1. `oura.yaml`
2. `spec.yaml`
3. schemas

## 🚨 User Impact 🚨

No changes to existing code.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [x] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

None.

</details>

<details><summary><strong>Integration</strong></summary>

image

</details>

<details><summary><strong>Acceptance</strong></summary>

image

</details>

### CODE REVIEW

**Comment by marcosmarxm** [comment]
Thanks @isaacharrisholt someone will review this during the week. Please ping me if no one review this during 48h :)

File: airbyte-integrations/connectors/source-oura/source_oura/spec.yaml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    start_datetime:
+      type: string
+      description: Start datetime for heart rate data in ISO 8601 format
+      pattern: ^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9]):([0-5][0-9])(Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9])?$
+      order: 3
+    end_datetime:
+      type: string
+      description: End datetime for heart rate data in ISO 8601 format
+      pattern: ^(-?(?:[1-9][0-9]*)?[0-9]{4})-(1[0-2]|0[1-9])-(3[01]|0[1-9]|[12][0-9])T(2[0-3]|[01][0-9]):([0-5][0-9]):([0-5][0-9])(Z|[+-](?:2[0-3]|[01][0-9]):[0-5][0-9])?$
+      order: 4
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-oura/source_oura/spec.yaml:34]: Definitely not a good pattern here. Please use string interpolation to extract the date from datetime.
Start Datetime is required?

File: airbyte-integrations/connectors/source-oura/source_oura/oura.yaml [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    http_method: "GET"
+    authenticator:
+      type: BearerAuthenticator
+      api_token: "{{ config['api_key'] }}"
+  date_requester:
+    $ref: "*ref(definitions.base_requester)"
+    request_options_provider:
+      request_parameters:
+        start_date: "{{ config['start_date'] }}"
+        end_date: "{{ config['end_date'] }}"
```
Comment (marcosmarxm) [inline airbyte-integrations/connectors/source-oura/source_oura/oura.yaml:18]:
```suggestion
        end_date: "{{ config['end_date'] or utc_now() }}"
```
Check more about macros here: https://github.com/airbytehq/airbyte/blob/master/airbyte-cdk/python/airbyte_cdk/sources/declarative/interpolation/macros.py

**Review by marcosmarxm — CHANGES_REQUESTED** [review]
A few comments about the config parameters.

File: airbyte-integrations/connectors/source-oura/source_oura/oura.yaml (hunk 2 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-oura/source_oura/oura.yaml:18]: This is the API default anyway, so I'm not sure this is necessary.

File: airbyte-integrations/connectors/source-oura/source_oura/spec.yaml (hunk 1 above)
Comment (isaacharrisholt) [reply airbyte-integrations/connectors/source-oura/source_oura/spec.yaml]: Good point. Have become more familiar with the low code CDK over the weekend, so much better with it now.

**Comment by marcosmarxm** [comment]
@isaacharrisholt please ask my review when you think all comments are solved.

**Comment by isaacharrisholt** [comment]
> @isaacharrisholt please ask my review when you think all comments are solved.

@marcosmarxm I think it's good to go

**Review by marcosmarxm — APPROVED** [review]
@isaacharrisholt please add the connector docs? I'll publish after that been addressed.

**Comment by isaacharrisholt** [comment]
@marcosmarxm docs added!

**Comment by marcosmarxm** [comment]
I tested locally the connector and it is working. Because this connector uses only personal data I didn't added the credentials in the CI.

**Pull request merged**